### PR TITLE
feat(web): allow a visitor to register a publisher account

### LINKED ISSUE #10: System allows a visitor to register a publisher account
## Description

A visitor must be able to register a game publisher account.

### PR DESCRIPTION
**Pull request by rllko**
Refs: #10 

## Summary

Add page to register a publisher account.

### CODE REVIEW

File: web/src/lib/api.ts [hunk 1]
```diff
@@ -165,6 +166,100 @@ export async function updateUser(id: string, details: EditableUser) {
   return user;
 }
 
+/**
+ * Creates a new publisher.
+ * @param newPublisher User to be created.
```
Comment (joaotomaspinheiro) [inline web/src/lib/api.ts]:
```suggestion
 * @param newPublisher Publisher to be created.
```

File: web/src/routes/distribute/register.tsx [hunk 2]
```diff
@@ -50,12 +56,20 @@ const formSchema = z
     vat: z.string().min(1, {
       message: "VAT is required",
     }),
-    picture: z.instanceof(File, {
-      message: "Picture is required",
-    }),
-    password: z.string().min(8, {
-      message: "Password must be at least 8 characters long",
-    }),
+    picture: z
```
Comment (joaotomaspinheiro) [inline web/src/routes/distribute/register.tsx]: Picture can be remove since it's not used in the form.

File: web/src/lib/api.ts [hunk 3]
```diff
@@ -165,6 +166,100 @@ export async function updateUser(id: string, details: EditableUser) {
...
+ * @returns a NewPublisher instance.
```
Comment (joaotomaspinheiro) [inline web/src/lib/api.ts]:
```suggestion
 * @returns Publisher created.
```

File: web/src/lib/api.ts [hunk 4]
```diff
@@ -165,6 +166,100 @@ export async function updateUser(id: string, details: EditableUser) {
...
+        throw new InternalServerError();
+    }
+  }
+
+  const jwt = (await response.json()) as Jwt;
+
+  return jwt;
+}
+
+export async function getPublisher(id: string) {
```
Comment (joaotomaspinheiro) [inline web/src/lib/api.ts]: This function is not used anywhere, so it can be removed.

File: web/src/lib/api.ts [hunk 5]
```diff
@@ -165,6 +166,100 @@ export async function updateUser(id: string, details: EditableUser) {
...
+      case 409:
+        throw new Conflict(error.code, error.message);
+      default:
+        throw new InternalServerError();
+    }
+  }
+
+  const user = (await response.json()) as User;
+
+  return user;
```
Comment (joaotomaspinheiro) [inline web/src/lib/api.ts]:
```suggestion
  const publisher = (await response.json()) as Publisher;

  return publisher;
```

File: web/src/domain/publisher.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  country?: string;
+  vatin?: string;
+  pictureMultimediaId?: string;
+}
+
+/**
+ * Represents the publisher credentials to sign in.
+ */
+export interface PublisherCredentials {
+  email?: string;
```
Comment (joaotomaspinheiro) [inline web/src/domain/publisher.ts]: Email is a required credential.

File: web/src/domain/publisher.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  address: string;
+  country: string;
+  vatin: string;
+  pictureMultimediaId?: string;
+}
+
+/**
+ * Represents an editable Publisher/PublisherPatch.
+ */
+export interface EditablePublisher {
```
Comment (joaotomaspinheiro) [inline web/src/domain/publisher.ts]: This is not used in this PR so it can be removed.

File: web/src/domain/publisher.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  address: string;
+  country: string;
+  vatin: string;
+  pictureMultimedia?: Multimedia;
+  createdAt: string;
+  modifiedAt: string;
+}
+
+/**
+ * Represents a new NewPublisher/PublisherPost.
```
Comment (joaotomaspinheiro) [inline web/src/domain/publisher.ts]:
```suggestion
 * Represents a new publisher.
```

File: web/src/domain/publisher.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
+import { Multimedia } from "./multimedia";
+
+/**
+ * Represents a Publisher.
```
Comment (joaotomaspinheiro) [inline web/src/domain/publisher.ts]:
```suggestion
 * Represents a publisher.
```

File: web/src/lib/api.ts [hunk 10]
```diff
@@ -165,6 +166,100 @@ export async function updateUser(id: string, details: EditableUser) {
   return user;
 }
 
+/**
+ * Creates a new publisher.
+ * @param newPublisher User to be created.
+ * @returns a NewPublisher instance.
+ * @throws {Conflict} Username, email or vatin already exist.
```
Comment (joaotomaspinheiro) [inline web/src/lib/api.ts]:
```suggestion
 * @throws {Conflict} Email or vatin already exist.
```

File: web/src/routes/distribute/register.tsx [hunk 11]
```diff
@@ -80,7 +96,58 @@ function Component() {
...
+      const payload = decodeTokenPayload(jwt.token);
+      storeToken(jwt.token, payload.exp);
+    },
+    async onSuccess() {
+      await navigate({ to: "/distribute/games" });
+    },
+    onError(error) {
+      if (error instanceof Conflict) {
+        switch (error.code) {
+          case "publisher_name_already_exists":
```
Comment (joaotomaspinheiro) [inline web/src/routes/distribute/register.tsx]: This case shouldn't exist.

File: web/src/lib/api.ts [hunk 12]
```diff
@@ -165,6 +170,100 @@ export async function updateUser(id: string, details: EditableUser) {
...
+    }
+  }
+
+  const publisher = (await response.json()) as Publisher;
+
+  return publisher;
+}
+
+/**
+ * Signs in a user with their credentials.
```
Comment (joaotomaspinheiro) [inline web/src/lib/api.ts]:
```suggestion
 * Signs in a publisher with their credentials.
```

File: web/src/lib/api.ts [hunk 13]
```diff
@@ -165,6 +170,100 @@ export async function updateUser(id: string, details: EditableUser) {
...
+ * @param credentials User credentials.
```
Comment (joaotomaspinheiro) [inline web/src/lib/api.ts]:
```suggestion
 * @param credentials Publisher credentials.
```

File: web/src/lib/api.ts [hunk 14]
```diff
@@ -165,6 +170,100 @@ export async function updateUser(id: string, details: EditableUser) {
...
+ * @returns JWT.
+ * @throws {Unauthorized} Incorrect credentials.
+ * @throws {InternalServerError} Server internal error.
+ */
+export async function signInPublisher(credentials: PublisherCredentials) {
+  const response = await fetch("/api/publishers/signin", {
+    signal: AbortSignal.timeout(DEFAULT_TIMEOUT),
+    method: "POST",
+    credentials: "include",
```
Comment (joaotomaspinheiro) [inline web/src/lib/api.ts]: This is not needed.

File: web/src/routes/distribute/register.tsx [hunk 15]
```diff
@@ -74,13 +85,59 @@ function Component() {
...
+
+      toast({
+        variant: "destructive",
+        title: "Oops! An unexpected error occurred",
+        description: "Please try again later or contact the support team.",
+      });
+    },
+  });
+
+  function onSubmit(data: RegisterSchemaType) {
```
Comment (joaotomaspinheiro) [inline web/src/routes/distribute/register.tsx:144]:
```suggestion
  /**
   * Handles form submission.
   * @param data Form data.
   */
  function onSubmit(data: RegisterSchemaType) {
```

File: web/src/routes/distribute/register.tsx [hunk 16]
```diff
@@ -74,13 +85,63 @@ function Component() {
...
+      const payload = decodeTokenPayload(jwt.token);
+      storeToken(jwt.token, payload.exp);
+    },
+    async onSuccess() {
+      await navigate({ to: "/distribute/games" });
+    },
+    onError(error) {
+      if (error instanceof Conflict) {
+        switch (error.code) {
+          case "user_email_already_exists":
```
Comment (Goncalo-Marques) [inline web/src/routes/distribute/register.tsx]:
```suggestion
          case "publisher_email_already_exists":
```

File: web/src/routes/distribute/register.tsx [hunk 17]
```diff
@@ -74,13 +85,63 @@ function Component() {
...
+            form.setError("vat", { message: "VAT already exists" });
+            break;
+        }
+        return;
+      }
+
+      toast({
+        variant: "destructive",
+        title: "Oops! An unexpected error occurred",
+        description: "Please try again later or contact the support team.",
```
Comment (Goncalo-Marques) [inline web/src/routes/distribute/register.tsx]: There is a const with this toast to avoid duplicated code

**Review by Goncalo-Marques — APPROVED** [review]

**Review by Goncalo-Marques — CHANGES_REQUESTED** [review]
I think that @joaotomaspinheiro still needs to perform some changes regarding the page redirects when a publisher is not signed in

**Review by Goncalo-Marques — APPROVED** [review]

**Review by joaotomaspinheiro — APPROVED** [review]

**Pull request merged**